src/util/threadpool.rs

```diff
@@ -254,115 +161,89 @@ where
 }
 
 // Each thread has a worker.
-struct Worker<Q, T> {
-    task_pool: Arc<(Mutex<TaskPool<Q, T>>, Condvar)>,
+struct Worker<C> {
+    stop_flag: Arc<AtomicBool>,
+    task_pool: Arc<(Mutex<FifoQueue<C>>, Condvar)>,
     task_count: Arc<AtomicUsize>,
+    ctx: C,
 }
 
-impl<Q, T> Worker<Q, T>
+impl<C> Worker<C>
 where
-    Q: ScheduleQueue<T>,
-    T: Debug,
+    C: Context,
 {
     fn new(
-        task_pool: Arc<(Mutex<TaskPool<Q, T>>, Condvar)>,
+        task_queue: Arc<(Mutex<FifoQueue<C>>, Condvar)>,
         task_count: Arc<AtomicUsize>,
-    ) -> Worker<Q, T> {
+        stop_flag: Arc<AtomicBool>,
+        ctx: C,
+    ) -> Worker<C> {
         Worker {
-            task_pool: task_pool,
+            stop_flag: stop_flag,
+            task_pool: task_queue,
             task_count: task_count,
+            ctx: ctx,
         }
     }
 
-    // `get_next_task` return `None` when `task_pool` is stopped.
-    #[inline]
-    fn get_next_task(&self, prev_gid: Option<&T>) -> Option<Task<T>> {
-        // try to receive notification.
+    fn get_task_timeout(&mut self, timeout: time::Duration) -> Option<Task<C>> {
         let &(ref lock, ref cvar) = &*self.task_pool;
         let mut task_pool = lock.lock().unwrap();
-        if prev_gid.is_some() {
-            task_pool.on_task_finished(prev_gid.unwrap());
+
+        if let Some(task) = (*task_pool).pop() {
+            return Some(task);
         }
-        loop {
-            if task_pool.is_stopped() {
-                return None;
-            }
-            if let Some(task) = task_pool.pop_task() {
-                // `on_task_started` should be here since:
-                //  1. To reduce lock's time;
-                //  2. For some schedula_queue,on_task_started should be
-                //  in the same lock with `pop_task` for the thread safety.
-                task_pool.on_task_started(&task.gid);
-                return Some(task);
-            }
-            task_pool = cvar.wait(task_pool).unwrap();
+        let (mut task_pool, _) = cvar.wait_timeout(task_pool, timeout).unwrap();
+        if let Some(task) = (*task_pool).pop() {
+            return Some(task);
         }
+        None
     }
 
     fn run(&mut self) {
-        let mut task = self.get_next_task(None);
-        // Start the worker. Loop breaks when receive stop message.
-        while let Some(t) = task {
-            // Since tikv would be down when any panic happens,
-            // we don't need to process panic case here.
-            (t.task)();
-            self.task_count.fetch_sub(1, AtomicOrdering::SeqCst);
-            task = self.get_next_task(Some(&t.gid));
+        while !self.stop_flag.load(AtomicOrdering::SeqCst) {
+            if let Some(t) = self.get_task_timeout(time::Duration::from_millis(WORKER_WAIT_TIME)) {
+                self.ctx.on_task_started();
```

Comment:
Does it better to provide a `on_tick` and run it each loop rather than checking time and `on_interval`? I think time check logic should be implement by user in `on_tick`.